server/sonar-process-monitor/src/main/java/org/sonar/application/process/CommandFactoryImpl.java

```diff
@@ -44,20 +50,68 @@
 
   private final AppSettings settings;
 
-  public JavaCommandFactoryImpl(AppSettings settings) {
+  public CommandFactoryImpl(AppSettings settings) {
     this.settings = settings;
   }
 
   @Override
-  public JavaCommand createEsCommand() {
-    File homeDir = settings.getProps().nonNullValueAsFile(ProcessProperties.PATH_HOME);
-    return newJavaCommand(ProcessId.ELASTICSEARCH, homeDir)
-      .addJavaOptions("-Djava.awt.headless=true")
-      .addJavaOptions(settings.getProps().nonNullValue(ProcessProperties.SEARCH_JAVA_OPTS))
-      .addJavaOptions(settings.getProps().nonNullValue(ProcessProperties.SEARCH_JAVA_ADDITIONAL_OPTS))
-      .setClassName("org.sonar.search.SearchServer")
-      .addClasspath("./lib/common/*")
-      .addClasspath("./lib/search/*");
+  public EsCommand createEsCommand(AppSettings settings) {
+    File homeDir = this.settings.getProps().nonNullValueAsFile(ProcessProperties.PATH_HOME);
+    File executable = new File(homeDir, getExecutable());
+    if (!executable.exists()) {
+      throw new IllegalStateException("Cannot find elasticsearch binary");
+    }
+
+    Map<String, String> settingsMap = new EsSettings(this.settings.getProps()).build();
+
+    File logDir = new File(settingsMap.get("path.logs"));
+    File confDir = new File(settingsMap.get("path.conf"));
+    Path jvmOptionsFile = confDir.toPath().resolve("jvm.options");
+    EsCommand res = new EsCommand(ProcessId.ELASTICSEARCH)
+      .setWorkDir(executable.getParentFile().getParentFile())
+      .setExecutable(executable)
+      .setConfDir(confDir)
+      .setLog4j2Properties(buildLog4j2Properties(logDir))
+      .setArguments(this.settings.getProps().rawProperties())
+      .setClusterName(settingsMap.get("cluster.name"))
+      .setHost(settingsMap.get("network.host"))
+      .setPort(Integer.valueOf(settingsMap.get("transport.tcp.port")))
+      .addJvmOption(settings.getProps().nonNullValue(ProcessProperties.SEARCH_JAVA_OPTS))
+      .addJvmOption(settings.getProps().nonNullValue(ProcessProperties.SEARCH_JAVA_ADDITIONAL_OPTS))
+      .setJvmOptionsFile(jvmOptionsFile)
+      .setEnvVariable("ES_JVM_OPTIONS", jvmOptionsFile.toString())
+      ;
+
+    settingsMap.forEach((key, value) -> res.addEsOption("-E" + key + "=" + value));
+
+    return res;
+  }
+
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)